harvdev_utils/chado_functions/cvterm.py

```diff
@@ -25,3 +35,89 @@ def get_cvterm(session, cv_name, cvterm_name):
         raise CodingError("HarvdevError: Could not find cv {}, cvterm {}.".format(cv_name, cvterm_name))
         return None
     return cv_cvterm[cv_name][cvterm_name]
+
+########################
+# cvterm props functions
+########################
+
+
+def check_cvterm_has_prop(session, cvterm, prop_value):
+    """Check cvterm has a specific prop value.
+
+    cvterm: (Cvterm Object) - Cvterm object.
+    prop_value:  (str) - Value fo cvterm prop.
+
+    Return True or False depending on wether it was found or not.
+    """
+    global cvterm_id_to_props
+    found = False
+    cvterm_id = cvterm.cvterm_id
+    if cvterm.cvterm_id in cvterm_id_to_props:
+        if prop_value in cvterm_id_to_props[cvterm_id]:
+            found = True
+        return found
+
+    # look up cvtermprops for this cvterm
+    props = session.query(Cvtermprop).filter(Cvtermprop.cvterm_id == cvterm_id).all()
+    cvterm_id_to_props[cvterm_id] = set()
+    for prop in props:
+        cvterm_id_to_props[cvterm_id].add(prop.value)
+    return prop_value in cvterm_id_to_props[cvterm_id]
+
+
+def check_cvterm_is_allowed(session, cvterm, list_of_props, retain_name=None):
+    """Check if cvterm is allowed.
+
+    cvterm: (Cvterm Object) - Cvterm object.
+    list_of_props: (list) list of db:propnames to lookup
+        i.e. ['FBdv:default', 'FBcv:environment_qualifier']
+    retain_name: <optional> (str)
+        If set will create and keep list of cvterms allowed and store this in retained.
+        NOTE: Will suck up memory depending on number BUT
+            will allow faster lookup later on if we anticipate
+            this being used a lot.
+        If not set it will create a name by joining list element into a str.
+
+    Raise Exception CodingError: if each one of the list of props does not fit the xxxx:bbbbbbb format
+                                 or does not find any cvterms (Useless addition).
+    """
+    global db_propname_to_cvterm_ids, retained
+
+    # create retained values for list of props
+    if not retain_name:
+        retain_name = '-'.join(list_of_props)
+    if retain_name in retained:
+        if cvterm.cvterm_id in retained[retain_name]:
+            return True
+        return False
```

Comment:
Not really, in fact some flake8 type messages may get flagged up as it might say 'useless use of else".
Actually not sure here. I think I do it this way as it may be faster it does not have to calculate the else bit???
But yes I can see it could be confusing.